src/main/java/graphql/schema/SchemaUtil.java

```diff
@@ -2,172 +2,86 @@
 
 
 import graphql.Assert;
-import graphql.AssertException;
 import graphql.Internal;
 import graphql.introspection.Introspection;
+import graphql.util.TraversalControl;
+import static graphql.util.TraversalControl.QUIT;
+import static graphql.schema.GraphqlTypeVisitors.LeafVisitor;
+import static graphql.schema.GraphqlTypeVisitors.CollectingVisitor;
+import static graphql.schema.GraphqlTypeVisitors.TypeResolvingVisitor;
+import graphql.util.TraverserContext;
 
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.HashMap;
-import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 
-import static java.lang.String.format;
 
 @Internal
 public class SchemaUtil {
 
-    public boolean isLeafType(GraphQLType type) {
-        GraphQLUnmodifiedType unmodifiedType = getUnmodifiedType(type);
-        return
-                unmodifiedType instanceof GraphQLScalarType
-                        || unmodifiedType instanceof GraphQLEnumType;
+    private static final TypeTraverser SHALLOW_TRAVERSER = new TypeTraverser((node) -> Collections.emptyList());
```

Comment:
I agree with @bbakerman: the original code for `isLeafType` and `isInputType` is so much simpler and I don't really see a non theoretical argument to replace that. So please revert it to the original code and delete the LeafVisitor.